README.md
<div align="center">

# stackage-parse

[![GitHub release (latest SemVer)](https://img.shields.io/github/v/release/tbidne/stackage-parse?include_prereleases&sort=semver)](https://github.com/tbidne/stackage-parse/releases/)
![haskell](https://img.shields.io/static/v1?label=&message=9.4&logo=haskell&logoColor=655889&labelColor=2f353e&color=655889)
[![MIT](https://img.shields.io/github/license/tbidne/stackage-parse?color=blue)](https://opensource.org/licenses/MIT)

[![nix](http://img.shields.io/github/actions/workflow/status/tbidne/stackage-parse/nix.yaml?branch=main&label=nix&logo=nixos&logoColor=85c5e7&labelColor=2f353c)](https://github.com/tbidne/stackage-parse/actions/workflows/nix.yaml)
[![cabal](http://img.shields.io/github/actions/workflow/status/tbidne/stackage-parse/cabal.yaml?branch=main&label=cabal&labelColor=2f353c)](https://github.com/tbidne/stackage-parse/actions/workflows/cabal.yaml)
[![style](http://img.shields.io/github/actions/workflow/status/tbidne/stackage-parse/style.yaml?branch=main&label=style&logoColor=white&labelColor=2f353c)](https://github.com/tbidne/stackage-parse/actions/workflows/style.yaml)

</div>

---

# Introduction

`stackage-parse` is a CLI utility for retrieving `stackage` snapshot data from [stackage.org](www.stackage.org). The primary motivation is testing with [clc-stackage](https://github.com/Bodigrim/clc-stackage): that is, easily updating a cabal file with every package from the latest stackage snapshot. See the [Pkgs](#pkgs) command for this specific workflow.

**Usage:**

```
stackage-parse: A tool for parsing stackage snapshot data.

Usage: stackage-parse [--lts (latest|LTS_STR)] [--nightly (latest|DATE_STR)]
                      [-e|--exclusions PATH] [-i|--inclusions PATH] COMMAND
                      [-v|--version]

Available options:
  --lts (latest|LTS_STR)   LTS snapshot e.g. 20.14 or the string 'latest'.
                           Overridden by --nightly.
  --nightly (latest|DATE_STR)
                           Nightly snapshot e.g. 2023-03-14 or the string
                           'latest'. Overrides --lts.
  -e,--exclusions PATH     Path to file with a list of packages to exclude from
                           the package list. Each package should be listed on a
                           separate line, without version numbers.
  -i,--inclusions PATH     Path to file with a list of packages to include from
                           the package list. Each package should be listed on a
                           separate line, without version numbers.
  -h,--help                Show this help text

Available commands:
  full                     Prints full package list and snapshot metadata
                           formatted as json.
  pkgs                     Lists all packages in a given snapshot.
  snapshot                 Prints snapshot metadata formatted as json.

Version: 0.1
```

# Commands

## Pkgs

**Description:** Lists all packages in a given snapshot. The package list is extremely long, so the examples here are truncated.

**Usage:**

```
Usage: stackage-parse pkgs [-f|--format (short|cabal)]
                           [-c|--comma (append|prepend)]

  Lists all packages in a given snapshot.

Available options:
  -f,--format (short|cabal)
                           Short corresponds to <pkg-name>-<vers> e.g.
                           'text-2.0.1'.Cabal corresponds to format suitable to
                           be pasted into a cabal file's 'build-depends' e.g.
                           'text ==2.0.1'. Defaults to short.
  -c,--comma (append|prepend)
                           If given, prepends/append a comma before/after each
                           entry.
  -h,--help                Show this help text
```

**Examples:**

As producing cabal-compatible stackage output is the main motivation, let us get right down to it:

```
$ stackage-parse pkgs -f cabal -c append
abstract-deque ==0.3,
abstract-deque-tests ==0.3,
abstract-par ==0.3.3,
AC-Angle ==1.0,
acc ==0.2.0.1,
...
```

* `pkgs` is the command.
* `-f cabal` outputs in `cabal` format i.e. `abstract-par ==0.3.3` instead of `abstract-par-0.3.3`.
* `-c append` appends a comma to each line.

This will produce a list of every package in the specified stackage snapshot (`stackage nightly` in this example). Because we may not want to include _all_ of stackage (e.g. executables, packages relying on external C libs), there are two ways to refine this list.

### Exclusions

```
$ stackage-parse pkgs --exclusions exclude-file -f cabal -c append
abstract-deque ==0.3,
...
```

Same as before, except any packages in `exclude-file` will be excluded from the package list. We can use this to exclude packages we **know** we do not want (e.g. because they will not build with our package).

### Inclusions

```
$ stackage-parse pkgs --inclusions include-file -f cabal -c append
abstract-deque ==0.3,
...
```

Like `--exclusions`, except only packages _in_ `include-file` will be included in the package list. We can use this to, say, upgrade only the packages from a previous snapshot without adding any new, potentially unbuildable packages.

See [examples](examples/) for more details.

## Snapshot

**Description:** Prints snapshot metadata as `json`. Here we use the `jq` utility to pretty-print the result.

**Usage:**
```
Usage: stackage-parse snapshot

  Prints snapshot metadata formatted as json.

Available options:
  -h,--help                Show this help text
```

**Examples:**

```
$ stackage-parse snapshot | jq
{
  "compiler": "ghc-9.4.4",
  "created": "2023-03-14",
  "ghc": "9.4.4",
  "name": "nightly-2023-03-14"
}
```

## Full

**Description:** Returns the package list and metadata as json i.e. combined `pkgs` and `snapshot` commands.

**Usage:**

```
Usage: stackage-parse full

  Prints full package list and snapshot metadata formatted as json.

Available options:
  -h,--help                Show this help text
```

**Examples:**

```
$ stackage-parse --nightly 2023-03-14 full | jq
{
  "packages": [
    <extremely long list omitted>
  ],
  "snapshot": {
    "compiler": "ghc-9.4.4",
    "created": "2023-03-14",
    "ghc": "9.4.4",
    "name": "nightly-2023-03-14"
  }
}
```

CHANGELOG.md
# Revision history for stackage-parse

## 0.1 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/Stackage/Args.hs
-- | Provides CLI args.
--
-- @since 0.1
module Stackage.Args
  ( getArgs,
    Args (..),
    Command (..),
    PkgListFormat (..),
    Comma (..),
  )
where

import Control.Applicative qualified as A
import Data.List qualified as L
import Data.String (IsString)
import Data.Text (Text)
import Data.Text qualified as T
import Options.Applicative
  ( CommandFields,
    InfoMod,
    Mod,
    OptionFields,
    Parser,
    ParserInfo (..),
    (<**>),
  )
import Options.Applicative qualified as OA
import Options.Applicative.Help.Chunk (Chunk (..))
import Options.Applicative.Types (ArgPolicy (..))
import Stackage.Data.Request
  ( SnapshotReq,
    mkSnapshotReqLatestLts,
    mkSnapshotReqLatestNightly,
    mkSnapshotReqLts,
    mkSnapshotReqNightly,
  )

-- | Retrieves CLI 'Args'.
--
-- @since 0.1
getArgs :: IO Args
getArgs = OA.execParser parserInfoArgs

-- | 'ParserInfo' type for parsing 'Args'.
--
-- @since 0.1
parserInfoArgs :: ParserInfo Args
parserInfoArgs =
  ParserInfo
    { infoParser = argsParser,
      infoFullDesc = True,
      infoProgDesc = Chunk desc,
      infoHeader = Chunk headerTxt,
      infoFooter = Chunk footerTxt,
      infoFailureCode = 1,
      infoPolicy = Intersperse
    }
  where
    headerTxt = Just "stackage-parse: A tool for parsing stackage snapshot data."
    footerTxt = Just versNum
    desc = Nothing

-- | CLI args.
--
-- @since 0.1
data Args = MkArgs
  { -- | @since 0.1
    ltsSnapshot :: !(Maybe SnapshotReq),
    -- | @since 0.1
    nightlySnapshot :: !(Maybe SnapshotReq),
    -- | @since 0.1
    exclusionsFile :: !(Maybe FilePath),
    -- | @since 0.1
    inclusionsFile :: !(Maybe FilePath),
    -- | @since 0.1
    command :: !Command
  }
  deriving stock
    ( -- | @since 0.1
      Eq,
      -- | @since 0.1
      Show
    )

-- | Format for package list.
--
-- @since 0.1
data PkgListFormat
  = -- | @since 0.1
    PkgListShort
  | -- | @since 0.1
    PkgListCabal
  deriving stock
    ( -- | @since 0.1
      Eq,
      -- | @since 0.1
      Show
    )

-- | Appends/Prepends a comma to the package list.
--
-- @since 0.1
data Comma
  = -- | @since 0.1
    CommaAppend
  | -- | @since 0.1
    CommaPrepend
  deriving stock
    ( -- | @since 0.1
      Eq,
      -- | @since 0.1
      Show
    )

-- | Command.
--
-- @since 0.1
data Command
  = -- | @since 0.1
    Full
  | -- | @since 0.1
    ListPackages !(Maybe PkgListFormat) !(Maybe Comma)
  | -- | @since 0.1
    GetSnapshot
  deriving stock
    ( -- | @since 0.1
      Eq,
      -- | @since 0.1
      Show
    )

argsParser :: Parser Args
argsParser =
  MkArgs
    <$> ltsSnapshotParser
    <*> nightlySnapshotParser
    <*> exclusionsFileParser
    <*> inclusionsFileParser
    <*> commandParser
    <**> OA.helper
    <**> version

ltsSnapshotParser :: Parser (Maybe SnapshotReq)
ltsSnapshotParser =
  snapshotParser' mkSnapshotReqLatestLts mkSnapshotReqLts options
  where
    options =
      mconcat
        [ OA.long "lts",
          OA.metavar "(latest|LTS_STR)",
          OA.help helpTxt
        ]
    helpTxt =
      mconcat
        [ "LTS snapshot e.g. 20.14 or the string 'latest'. ",
          "Overridden by --nightly."
        ]

nightlySnapshotParser :: Parser (Maybe SnapshotReq)
nightlySnapshotParser =
  snapshotParser' mkSnapshotReqLatestNightly mkSnapshotReqNightly options
  where
    options =
      mconcat
        [ OA.long "nightly",
          OA.metavar "(latest|DATE_STR)",
          OA.help helpTxt
        ]
    helpTxt =
      mconcat
        [ "Nightly snapshot e.g. 2023-03-14 or the string 'latest'. ",
          "Overrides --lts."
        ]

exclusionsFileParser :: Parser (Maybe FilePath)
exclusionsFileParser =
  A.optional $
    OA.option OA.str $
      mconcat
        [ OA.long "exclusions",
          OA.short 'e',
          OA.metavar "PATH",
          OA.help helpTxt
        ]
  where
    helpTxt =
      mconcat
        [ "Path to file with a list of packages to exclude from the package ",
          "list. Each package should be listed on a separate line, without ",
          "version numbers e.g. text."
        ]

inclusionsFileParser :: Parser (Maybe FilePath)
inclusionsFileParser =
  A.optional $
    OA.option OA.str $
      mconcat
        [ OA.long "inclusions",
          OA.short 'i',
          OA.metavar "PATH",
          OA.help helpTxt
        ]
  where
    helpTxt =
      mconcat
        [ "Path to file with a list of packages to include from the package ",
          "list. Each package should be listed on a separate line, without ",
          "version numbers e.g. text."
        ]

snapshotParser' ::
  -- | Constructor for latest snapshot.
  SnapshotReq ->
  -- | Constructor for specific snapshot. Can fail.
  (Text -> Either Text SnapshotReq) ->
  -- | Options.
  Mod OptionFields SnapshotReq ->
  Parser (Maybe SnapshotReq)
snapshotParser' mkLatest mk = A.optional . OA.option readSnapshot
  where
    readSnapshot =
      OA.str >>= \case
        "latest" -> pure mkLatest
        other -> case mk other of
          Right s -> pure s
          Left err -> fail $ T.unpack err

commandParser :: Parser Command
commandParser =
  OA.hsubparser $
    mconcat
      [ mkCommand "full" fullParser fullHelp,
        mkCommand "pkgs" listPackagesParser listPackagesHelp,
        mkCommand "snapshot" getSnapshotParser getSnapshotHelp
      ]
  where
    fullParser = pure Full
    fullHelp =
      OA.progDesc
        "Prints full package list and snapshot metadata formatted as json."

    listPackagesParser = ListPackages <$> pkgListFormatParser <*> commaParser
    listPackagesHelp = OA.progDesc "Lists all packages in a given snapshot."

    getSnapshotParser = pure GetSnapshot
    getSnapshotHelp = OA.progDesc "Prints snapshot metadata formatted as json."

pkgListFormatParser :: Parser (Maybe PkgListFormat)
pkgListFormatParser =
  A.optional $
    OA.option readListFormat $
      mconcat
        [ OA.long "format",
          OA.short 'f',
          OA.metavar "(short|cabal)",
          OA.help helpTxt
        ]
  where
    readListFormat =
      OA.str @String >>= \case
        "short" -> pure PkgListShort
        "cabal" -> pure PkgListCabal
        other -> fail $ "Unknown format: " ++ other
    helpTxt =
      mconcat
        [ "Short corresponds to <pkg-name>-<vers> e.g. 'text-2.0.1'.",
          "Cabal corresponds to format suitable to be pasted into a cabal ",
          "file's 'build-depends' e.g. 'text ==2.0.1'. Defaults to short."
        ]

commaParser :: Parser (Maybe Comma)
commaParser =
  A.optional $
    OA.option readComma $
      mconcat
        [ OA.long "comma",
          OA.short 'c',
          OA.metavar "(append|prepend)",
          OA.help helpTxt
        ]
  where
    readComma =
      OA.str @String >>= \case
        "append" -> pure CommaAppend
        "prepend" -> pure CommaPrepend
        other -> fail $ "Unknown comma: " ++ other
    helpTxt = "If given, prepends/appends a comma before/after each entry."

mkCommand :: String -> Parser a -> InfoMod a -> Mod CommandFields a
mkCommand cmdTxt parser helpTxt = OA.command cmdTxt (OA.info parser helpTxt)

version :: Parser (a -> a)
version = OA.infoOption txt (OA.long "version" <> OA.short 'v')
  where
    txt =
      L.intercalate
        "\n"
        [ "stackage-parse",
          versNum
        ]

versNum :: (IsString a) => a
versNum = "Version: 0.1"
